Add optional mapping of keys to short arguments

mshenfield opened the pull request:
A generic solution to close #16. Allow the user to pass in
a `shortargs` object in the `options`. Each entry maps an
accepted value to a short argument. For example, passing

`dargs({'file': 'value'}, {shortargs: {'file': 'f'}})`

might produce an output like

`[ '-f value']`

This is nice because it allows you to work with some commands
that only support short arg versions of certain arguments (I
was working with `ctags -f`), while still encouraging you to
expose expressive, full word arguments to the end user.

Updated tests with usage alone and with `includes`, and docs
with examples and short description.

sindresorhus commented on index.js:
```diff
@@ -1,7 +1,11 @@
+function createArg(key, val, shortargs) {
+	if (typeof shortargs === 'object' && shortargs[key]) {
+		return '-' + shortargs[key] + (val ? ' ' + val : '')
+	}
```
It would be better to handle the conversion to shortform on line 28. Then this could only be responsible for handling one character options and we wouldn't have to pass the `shortargs` argument.

sindresorhus commented on readme.md:
```diff
@@ -92,6 +123,12 @@ Type: `array`
 
 Keys to include.
 
+##### shortargs
```
Needs a better name, and should be camelCased.

mshenfield replied:
Naming is my bane :grimacing:. Do any of these seem more expressive: `shortFlags`, `shortForms`, `singleDashAliases`?

sindresorhus commented on readme.md:
```diff
@@ -23,11 +23,13 @@ var input = {
 
 var excludes = ['sad'];
 var includes = ['camelCase', 'multiple', 'sad'];
+var shortargs = {'file', 'f'}
```
No need to quote `'file'`.

radkodinev replied:
And it must be with a colon instead of comma, i.e. `{ file: 'f' }`.

sindresorhus commented on test.js:
```diff
@@ -74,6 +74,33 @@ test('excludes and includes options', function (t) {
 });
 
+test('shortargs options', function (t) {
+	var actual = dargs({'a': 'foo', 'file': 'test'}, {shortargs: {'file': 'f'}});
```
No need to quote `'file'`.

radkodinev commented on readme.md:
```diff
@@ -92,6 +123,12 @@ Type: `array`
+
+Type: `object`
+
+Maps keys to short arg replacements. Keys are still effected by `includes` and `excludes`.
```
effected or affected?

sindresorhus replied:
affected

mshenfield commented:
Updated the diff to correct "effected" to "affected", key quoting, change `shortargs` to `shortForms`, and add [`createShortFormArg`](https://github.com/mshenfield/dargs/commit/38d79f69dc674bcc9735316a73ab5df39fc77d05#diff-168726dbe96b3ce427e7fedce31bb0bcR9), separate function for generating a single "-" argument string for a key and a value.

I created a variable to represent [the function used to generate the argument string](https://github.com/mshenfield/dargs/commit/38d79f69dc674bcc9735316a73ab5df39fc77d05#diff-168726dbe96b3ce427e7fedce31bb0bcR20), which defaults to `createArg`.  If a key is found in the `shortForms` object, that variable is set to `createShortFormArg`. This works because both functions have the same signature, which may be fragile.

sindresorhus commented:
Can you fix the merge conflict?

sindresorhus commented:
Minimist uses the option name `alias`. Maybe we should too?

sindresorhus commented on readme.md:
```diff
@@ -63,6 +65,35 @@ console.log(dargs(input, {includes: includes}));
 */
+
+
+console.log(dargs(input, {shortForms: shortForms}));
```
To make it clearer what this does I think we should drop using the `input` fixture and just have one inline input here.

sindresorhus commented on readme.md:
```diff
@@ -63,6 +65,35 @@ console.log(dargs(input, {includes: includes}));
+	'--multiple=value2',
+	'--sad=:('
+]
+*/
```
I don't think we need this example. It's already noted in the API docs.

sindresorhus commented on test.js:
```diff
@@ -74,6 +74,33 @@ test('excludes and includes options', function (t) {
+		'-f test'
+	];
+	t.assert(deepEqual(actual, expected));
+	t.end();
```
Don't need this anymore.

sindresorhus commented on test.js:
```diff
@@ -74,6 +74,33 @@ test('excludes and includes options', function (t) {
+		'--camel-case-camel'
+	];
+	t.assert(deepEqual(actual, expected));
+	t.end();
```
Don't need this anymore.

sindresorhus commented on test.js:
```diff
@@ -74,6 +74,33 @@ test('excludes and includes options', function (t) {
+test('shortForms options', function (t) {
+	var actual = dargs({a: 'foo', file: 'test'}, {shortForms: {file: 'f'}});
+	var expected = [
+		'--a=foo',
```
This should be with one dash, as it should be possible to pass short forms directly too.

mshenfield replied:
Implicitly causing any single letter arguments to use a single "-" makes sense, but I think should be a separate pull request. There is actually an [existing closed pull request](https://github.com/sindresorhus/dargs/pull/4) that does just that. 

This is focused on making the conversion to single "-" arguments explicit so that the end user can handle cases where only the single "-" argument is available, but still expose a readable `options` input to be handled by `dargs`.

mshenfield commented:
Switched to `alias` and added a sentence to the documentation explaining that these arguments are created with a single "-" and a space separating the name form the value. Fixed merge conflicts, using `ava` framework for new tests. Added inline `input` and removed extra example in the readme.

sindresorhus commented on index.js:
```diff
@@ -19,6 +23,7 @@ module.exports = function (input, opts) {
 
 	Object.keys(input).forEach(function (key) {
 		var val = input[key];
+		var argFunction = createArg;
```
`var argFn = createArg;`

sindresorhus commented on readme.md:
```diff
@@ -29,6 +29,7 @@ const input = {
 
 const excludes = ['sad', /.*Kind$/];  // excludes and includes accept regular expressions
 const includes = ['camelCase', 'multiple', 'sad', /^pie.*/];
+const aliases = {file: 'f'}
```
semicolon

sindresorhus commented on readme.md:
```diff
@@ -61,8 +62,24 @@ console.log(dargs(input, {includes: includes}));
-	'--sad=:(',
-	'--pie-kind=cherry'
+	'--pie-kind=cherry',
+	'--sad=:('
```
You can't switch the order here without switching the `includes` array order too.

sindresorhus commented on readme.md:
```diff
@@ -94,6 +111,12 @@ Type: `array`
 
+##### aliases
+
+Type: `object`g
```
`g`?

sindresorhus commented on test.js:
```diff
@@ -63,3 +63,28 @@ test('excludes and includes options', t => {
 	t.same(fn({foo: false}, {ignoreFalse: true}), []);
 });
+
+test('aliases option', function (t) {
```
use arrow function like the other tests

sindresorhus commented on test.js:
```diff
@@ -63,3 +63,28 @@ test('excludes and includes options', t => {
+		]);
+});
+
+test('includes and aliases options', function (t) {
```
use arrow function like the other tests

mshenfield commented:
Removed extra "g" and added missing semicolon in readme, used arrow function in tests, and changed name of `argFunction ` to `argFn`.

sindresorhus commented:
Looks good. Nice work on this one @mshenfield :)

mshenfield commented:
Thanks!
